pipelines/google_sheets/google_sheets.py

```diff
@@ -0,0 +1,133 @@
+def _initialize_sheets(credentials: GcpClientCredentialsWithDefault) -> Any:
+    # Build the service object for Google sheets api.
+    service = build('sheets', 'v4', credentials=credentials.to_service_account_credentials())
+    return service
+
+
+@dlt.source
+def google_spreadsheet(spreadsheet_id: str = None, spreadsheet_url: str = None, sheet_names: list[str] = None, credentials: GcpClientCredentialsWithDefault = dlt.secrets.value) -> Any:
+
+    # authenticate to the service using the helper function
+    service = _initialize_sheets(cast(GcpClientCredentialsWithDefault, credentials))
+    sheet = service.spreadsheets()
+
+    # get the id from url if the id was left empty
+    if not spreadsheet_id:
+        spreadsheet_id = process_url(spreadsheet_url)
+
+    # if sheet names is left empty, make an api call and discover all the sheets in the spreadsheeet
+    if (not sheet_names) or len(sheet_names) == 0:
+        sheet_names = []
+
+        # get metadata of spreadsheet to check for number of sheets inside
+        metadata = sheet.get(spreadsheetId=spreadsheet_id).execute()
+        sheet_meta = metadata['sheets']
+        # iterate through the sheets in the metadata and get their names
+        for sheet_m in sheet_meta:
+            # get name and append to list of sheet names
+            sheet_name = sheet_m['properties']['title']
+            sheet_names.append(sheet_name)
+
+    # Make 2 api calls to get the data for all sheets and ranges
+    # Get values where dates are a string and get values where dates are a serial number, this is done so the fields which are dates can be distinguished and properly calculated
+    # get values as a string
+    values = sheet.values().batchGet(
+        spreadsheetId=spreadsheet_id,
+        ranges=sheet_names,
+        # unformatted returns typed values
+        valueRenderOption="UNFORMATTED_VALUE",
+        # will return formatted dates as string
+        dateTimeRenderOption="FORMATTED_STRING"
+    ).execute()['valueRanges']
+
+    # get dates as serial number
+    values_date_sn = sheet.values().batchGet(
+        spreadsheetId=spreadsheet_id,
+        ranges=sheet_names,
+        # unformatted returns typed values
+        valueRenderOption="UNFORMATTED_VALUE",
+        # will return formatted dates as a serial number
+        dateTimeRenderOption="SERIAL_NUMBER"
+    ).execute()['valueRanges']
+
+    # iterate through both sets of values, same indexes work
+    for i in range(len(values)):
+        # get range names and values
+        sheet_range_name = values[i]['range']
+        sheet_range = values[i]['values']
+        sheet_range_sn = values_date_sn[i]['values']
+
+        # create a resource from processing both sheet/range values
+        yield dlt.resource(process_range(sheet_val=sheet_range, sheet_val_sn=sheet_range_sn),
+                           name=sheet_range_name,
+                           write_disposition="replace")
+
+
+if __name__ == "__main__":
```

Comment:
see `chess_pipeline.py` for an example of how to create usage examples. you should not have `__main__` here